File: src/2024/day1.zig
const std = @import("std");
const print = std.debug.print;
const input = @embedFile("day1.txt");

pub fn part1(left: []i32, right: []i32) void {
    std.mem.sort(i32, left, {}, std.sort.asc(i32));
    std.mem.sort(i32, right, {}, std.sort.asc(i32));

    var sum: u32 = 0;
    for (0..left.len) |i| {
        sum += @abs(left[i] - right[i]);
    }

    print("part1: {d}\n", .{sum});
}

pub fn part2(left: []i32, h: *std.AutoHashMap(i32, i32)) void {
    var sum: i32 = 0;
    for (left) |i| {
        const count = h.get(i) orelse 0;
        sum += i * count;
    }
    print("part2: {d}\n", .{sum});
}

pub fn part2Mem(left: []i32, right: []i32) void {
    var sum: i32 = 0;
    for (left) |i| {
        const count: i32 = @intCast(std.mem.count(i32, right, &[_]i32{i}));
        sum += count * i;
    }
    print("part2 mem: {d}\n", .{sum});
}

pub fn solve() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();
    var h = std.AutoHashMap(i32, i32).init(allocator);
    defer h.deinit();

    var it = std.mem.tokenizeScalar(u8, input, '\n');

    var a: std.ArrayList(i32) = std.ArrayList(i32).init(allocator);
    var b: std.ArrayList(i32) = std.ArrayList(i32).init(allocator);

    while (it.next()) |token| {
        var line = std.mem.split(u8, token, "   ");
        const left_value = try std.fmt.parseInt(i32, line.next().?, 10);
        const right_value = try std.fmt.parseInt(i32, line.next().?, 10);
        try a.append(left_value);
        try b.append(right_value);
        if (h.get(right_value)) |v| try h.put(right_value, v + 1) else try h.put(right_value, 1);
    }

    const left = try a.toOwnedSlice();
    const right = try b.toOwnedSlice();

    part1(left, right);
    part2(left, &h);
    part2Mem(left, right);
}

File: src/2024/day1.rs
use std::collections::HashMap;
use std::fs;

fn part1(left: &mut [i32], right: &mut [i32]) {
    left.sort_unstable();
    right.sort_unstable();

    let sum: i32 = left
        .iter()
        .zip(right.iter())
        .map(|(l, r)| (l - r).abs())
        .sum();

    println!("part1: {}", sum);
}

fn part2(left: &[i32], counts: &HashMap<i32, i32>) {
    let sum: i32 = left
        .iter()
        .map(|&i| i * counts.get(&i).copied().unwrap_or(0))
        .sum();

    println!("part2: {}", sum);
}

fn solve() -> Result<(), Box<dyn std::error::Error>> {
    let input = fs::read_to_string("day1.txt")?;

    let mut counts: HashMap<i32, i32> = HashMap::new();
    let mut left = Vec::new();
    let mut right = Vec::new();

    for line in input.lines() {
        let parts: Vec<&str> = line.split("   ").collect();
        if parts.len() == 2 {
            let left_value: i32 = parts[0].trim().parse()?;
            let right_value: i32 = parts[1].trim().parse()?;

            left.push(left_value);
            right.push(right_value);

            *counts.entry(right_value).or_insert(0) += 1;
        }
    }

    part1(&mut left, &mut right);
    part2(&left, &counts);

    Ok(())
}

fn main() {
    if let Err(e) = solve() {
        eprintln!("Error: {}", e);
    }
}
